pkg/kv/kvpb/api.proto

```diff
@@ -2883,9 +2884,20 @@ message Header {
   // and/or been explicitly committed by a RecoverTxn request. See #103817.
   bool ambiguous_replay_protection = 32;
 
+  // ConnectionClass specifies the preferred RPC connection class to use for
+  // sending or retransmitting this BatchRequest.
+  //
+  // This is only a hint. Senders of this request may override this preference.
+  // For example, requests to some critical system ranges can be "upgraded" to
+  // the SYSTEM class.
+  //
+  // If the enum value is unknown (for instance, may happen when the cluster is
+  // in a mixed-version state after a new value got added), DEFAULT is assumed.
```

Comment:
The `DEFAULT` safety is not implemented yet, the values are passed as is.

Maybe we don't need to reset to `DEFAULT`, just assume that all values of this are acceptable. It's not safe though, to things that depend on [NumConnectionClasses](https://github.com/cockroachdb/cockroach/blob/8484d4bf6af7866955183fa7223b30084f511697/pkg/rpc/connection_class.go#L45) being greater than the max connection class.

Reply:
I'd like to avoid having migrations boilerplate later in case we want to add a new type.

Reply:
The downside of option (1) is that the initiator of the request has very little control of which class will end up being used on the retransmitting nodes.

Example: suppose we want to split off some portion of `RAFT` class into a subset (e.g. `RAFT_SNAPSHOTS`). We don't want nodes that don't know `RAFT_SNAPSHOT` yet to suddenly switch to the `DEFAULT` class. A good behaviour would be to default to the "old" `RAFT` in this case.

Option (3) gives best control of this, because the initiator of the request can set `RAFT` or `RAFT_SNAPSHOT` depending on the version gate state.

Option (2) is somewhat worse, but at least we know that the request will go through a separate connection.

Option (1) is worst, because it can break old assumptions about what goes where.

Reply:
Isn't this connection class business purely a client side specific construct and does not effect
the server side in any way and thus likely doesn't not have any mix-version case issues?
The client either knows that there is a new "RAFT_SNAPSHOT" class, or it doesn't -- and it's
the client that initiates the connections.

 I understand that (unfortunately) this information is part of kvpb header -- and so theoretically
this connection class could be used by the server side which would raise mix-version issues.
(I say unfortunately because, well.. it's unfortunate because this information is connection
specific and not KV request specific.  It is howerver, very convienient place and at least
somewhat similar in spirit to "admission" header stuff, so -- that's fine (imo)).

Is the above a concern (meaing the use of this connection class on the server side)?  
Could this header field be marked (in the comment) as "client side only"and just be done with it?
Can we punt this until we actually have a situation where the server cares about
connection class (perhaps because it tries to establish its own rpc connections?)

Ragardless, connection class, I feel is an optimization and is a best effort.
If we want the server to support "unknown" connection classes, then we need to have the header
use an "int" instead of strongly typed enum to represent connection class; and the code
that translates it to well defined connection classes, with a fallback to default  doesn't sound that bad to me.

Reply:
My understanding is that `Send(BatchRequest)` can be transparently retransmitted by other nodes. So there is no exactly one client-server here.

Quoting Erik:
> Consider e.g. the case of follower routing, where we'll proxy via some other node's `DistSender` -- we'll want it to use the same connection class.

Reply:
> If we want the server to support "unknown" connection classes, then we need to have the header
use an "int" instead of strongly typed enum to represent connection class; and the code
that translates it to well defined connection classes, with a fallback to default doesn't sound that bad to me.

I agree, also leaning towards not taking option (2). Using option (1) with assuming that it's possible we will need option (3) some day seems like a good compromise.

Reply:
Added a commit implementing this.